src/structures/paging/mapper/recursive_page_table.rs

```diff
@@ -286,6 +285,121 @@ impl<'a> RecursivePageTable<'a> {
+    ///
+    /// ## Safety
+    ///
+    /// The caller has to guarantee that it's safe to free page table frames:
+    /// All page table frames must only be used once and only in this page table
+    /// (e.g. no reference counted page tables or reusing the same page tables for different virtual addresses ranges in the same page table).
+    #[inline]
+    pub unsafe fn clean_up<D>(&mut self, frame_deallocator: &mut D)
+    where
+        D: FrameDeallocator<Size4KiB>,
+    {
+        self.clean_up_addr_range(
+            PageRangeInclusive {
+                start: Page::from_start_address(VirtAddr::new(0)).unwrap(),
+                end: Page::from_start_address(VirtAddr::new(0xffff_ffff_ffff_f000)).unwrap(),
+            },
+            frame_deallocator,
+        )
+    }
+
+    /// Remove all empty P1-P3 tables in a certain range
+    /// ```
+    /// # use core::ops::RangeInclusive;
+    /// # use x86_64::{VirtAddr, structures::paging::{
+    /// #    FrameDeallocator, Size4KiB, MappedPageTable, mapper::RecursivePageTable, page::{Page, PageRangeInclusive},
+    /// # }};
+    /// # unsafe fn test(page_table: &mut RecursivePageTable, frame_deallocator: &mut impl FrameDeallocator<Size4KiB>) {
+    /// // clean up all page tables in the lower half of the address space
+    /// let lower_half = Page::range_inclusive(
+    ///     Page::containing_address(VirtAddr::new(0)),
+    ///     Page::containing_address(VirtAddr::new(0x0000_7fff_ffff_ffff)),
+    /// );
+    /// page_table.clean_up_addr_range(lower_half, frame_deallocator);
+    /// # }
+    /// ```
+    ///
+    /// ## Safety
+    ///
+    /// The caller has to guarantee that it's safe to free page table frames:
+    /// All page table frames must only be used once and only in this page table
+    /// (e.g. no reference counted page tables or reusing the same page tables for different virtual addresses ranges in the same page table).
+    pub unsafe fn clean_up_addr_range<D>(
+        &mut self,
+        range: PageRangeInclusive,
+        frame_deallocator: &mut D,
+    ) where
+        D: FrameDeallocator<Size4KiB>,
+    {
+        fn clean_up(
+            recursive_index: PageTableIndex,
+            page_table: &mut PageTable,
+            level: u8,
+            range: PageRangeInclusive,
+            frame_deallocator: &mut impl FrameDeallocator<Size4KiB>,
+        ) -> bool {
+            if range.is_empty() {
+                return false;
+            }
+
+            let table_addr = range
+                .start
+                .start_address()
+                .align_down(1u64 << (level * 9 + 12));
+
+            let start = range.start.p_index(level);
+            let end = range.end.p_index(level);
+
+            let mut is_empty = true;
+
+            let offset_per_entry = 1u64 << (((level - 1) * 9) + 12);
+            for (i, entry) in page_table.iter_mut().enumerate() {
+                if usize::from(start) < i && i <= usize::from(end) && level != 1 {
```

Comment:
We should skip the recursive entry of the level 4 table here because it references the level 4 table itself. I'm not sure if anything bad would happen, but we would definitely do duplicated work because we would visit the level 4 table multiple times.

```suggestion
                if usize::from(start) < i && i <= usize::from(end) && level != 1 && !(level == 4 && i == recursive_index) {
```

Reply:
Good idea. I added it with by calling filter on the iterator as I got rid of that if block earlier anyways.